Add packetcapture api

hangyan opened the pull request:
Add packetcapture's API changes. 

For the full feature, please ref to: 

https://github.com/antrea-io/antrea/pull/5821

hangyan commented:
moved API to this MR. Thanks you previous feedback and welcome to help review this again.

@luolanzone @tnqn @jianjuns @antoninbas 

Accoding to https://github.com/antrea-io/antrea/pull/5821#discussion_r1575226623 . `IPHeader` field was removed.

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -291,3 +291,151 @@ type TLSProtocol struct {
 	SNI string `json:"sni,omitempty"`
 }
+
+// Source describes the source spec of the traceflow.
```
The references to "traceflow" (3 in this file) are probably not correct. It seems that eventually we may want to use the same structs for the `Traceflow` API and the `PacketCapture` API, but for now at least they don't share an API version and we should remove the references to "traceflow" here.

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -291,3 +291,151 @@ type TLSProtocol struct {
+// IPHeader describes spec of an IPv4 header.
+type IPHeader struct {
+	// Protocol is the IP protocol.
+	Protocol int32 `json:"protocol,omitempty" yaml:"protocol,omitempty"`
+	// TTL is the IP TTL.
+	TTL int32 `json:"ttl,omitempty" yaml:"ttl,omitempty"`
+	// Flags is the flags for IP.
+	Flags int32 `json:"flags,omitempty" yaml:"flags,omitempty"`
```
why do we have partial `yaml` tags (some structs have yaml tags for their fields, while others don't)?

hangyan replied:
I'm not sure why in the first place. I guess we can remove all the yaml tags.

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -291,3 +291,151 @@ type TLSProtocol struct {
+	DstIP string `json:"dstIP,omitempty"`
+	// Length is the IP packet length (includes the IPv4 or IPv6 header length).
+	Length          int32           `json:"length,omitempty"`
+	IPv6Header      *IPv6Header     `json:"ipv6Header,omitempty"`
```
I didn't follow the previous discussion too closely, but it it surprising that we have an `IPv6Header` field, but no corresponding field for IPv4.

hangyan replied:
There weren't any discussion about this before. I guess for ipv4 there weren't any important fields other than src or dst ip.

hangyan replied:
In Traceflow `v1beta1.Packet`, there is a `IPHeader` field. But in PacketCapture, we didn't have the implementation yet.  The filter support for TTL/flags didn't seems very important in the first place, so during the previous discussion, we have removed this field.

antoninbas replied:
So do we just want to be able to filter based on IP family, or do we care about filtering based on `NextHeader` / `HopLimit`?
This seems a bit inconsistent.
We also don't have a comment on the struct field to explain what it is used for.

hangyan replied:
(Hi , sorry for the absence before, from now on, i will continue to work on this actively to get it merged into v2.2 )

I feel like it would be better to remove `IPv6Header` and added a `IPFamily (string)` field would be enough. The `IPv6Header` struct is not used for filter.

The only problem is for the version upgrade case. Since `v1beta1.Traceflow` is using a similar struct `v1beta1.Packet`, i was hoping we could reuse this in the future. But i guess it won't be much a problem to keep it separated.

antoninbas replied:
We need to decide if we want these to be unified in the end or not. If not, then we just have to focus on PacketCapture here, but we may also have to add a unique prefix (`PacketCapture`) to all these structs. If we want them unified, maybe we should have a plan and see if it actually makes sense?

jianjuns replied:
I feel an IP protocol field should be useful. If we have that, maybe we can remove ICMP header?

Traceflow needs to generate a packet, which requires more fields about the packet. Maybe ok not to share packet struct between these two.

hangyan replied:
mark this as resloved as a new 'Protocol` field was added. we can continue to disscuss this in the new thread.

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -291,3 +291,151 @@ type TLSProtocol struct {
+}
+
+// IPHeader describes spec of an IPv4 header.
+type IPHeader struct {
```
unused?

hangyan replied:
removed

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -291,3 +291,151 @@ type TLSProtocol struct {
+// PacketCaptureFirstNConfig contains the config for the FirstN type capture. The only supported parameter is
+// `Number` at the moment, meaning capturing the first specified number of packets in a flow.
+type PacketCaptureFirstNConfig struct {
+	Number int32 `json:"number,omitempty"`
```
You should remove the `omitempty` if this is a required field

hangyan replied:
update

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -291,3 +291,151 @@ type TLSProtocol struct {
+	// Length is the IP packet length (includes the IPv4 or IPv6 header length).
+	Length          int32           `json:"length,omitempty"`
+	IPv6Header      *IPv6Header     `json:"ipv6Header,omitempty"`
+	TransportHeader TransportHeader `json:"transportHeader,omitempty"`
```
If this field is optional, let's make it a pointer. If it's not optional, make it a required property and let's remove the `omitempty` tag.

hangyan replied:
updated

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -291,3 +291,151 @@ type TLSProtocol struct {
+type Packet struct {
+	SrcIP string `json:"srcIP,omitempty"`
+	DstIP string `json:"dstIP,omitempty"`
+	// Length is the IP packet length (includes the IPv4 or IPv6 header length).
+	Length          int32           `json:"length,omitempty"`
```
is this really useful as a filter?

hangyan replied:
No. will remove the unused fields.

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -291,3 +291,151 @@ type TLSProtocol struct {
+
+// UDPHeader describes spec of a UDP header.
+type UDPHeader struct {
+	// SrcPort is the source port.
+	SrcPort int32 `json:"srcPort,omitempty"`
+	// DstPort is the destination port.
+	DstPort int32 `json:"dstPort,omitempty"`
```
I don't know how @jianjuns and @tnqn will feel about this, but since this is a new API, I think we should follow the [K8s convention](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required):

> Therefore, we ask that pointers always be used with optional fields that do not have a built-in nil value.

It tends to lead to a lot of pointers, but it makes it clearer what is optional and what is not. When creating a resource with YAML, it makes no difference. When creating a resource in Go, we can use `ptr.To` from the K8s libraries to make initialization less verbose. What do you think?

hangyan replied:
Agreed. There maybe a small problem if we upgrade to v1beta1 later to share the same structure with Traceflow though.

antoninbas replied:
I'd rather do the best interface possible now for `PacketCapture`, and worry about unifying later.

If we have to introduce a new `v1beta2` version for Traceflow, it should still be possible to define a conversion webhook to convert between `v1beta1` and `v1beta2`, because of the semantics of the field?

hangyan replied:
updated. Can you help review this change? I'm not very sure if i have done it right.

tnqn replied:
I think @antoninbas means:
```
type UDPHeader struct {
	// SrcPort is the source port.
	SrcPort *int32 `json:"srcPort,omitempty"`
	// DstPort is the destination port.
	DstPort *int32 `json:"dstPort,omitempty"`
}
```
because they are optional.

tnqn replied:
Just saw newer comments. Please ignore it.

luolanzone commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -291,3 +291,151 @@ type TLSProtocol struct {
+	// NumCapturedPackets records how many packets have been captured. If it reaches the target number, the capture
+	// can be considered as finished.
+	NumCapturedPackets int32 `json:"numCapturedPackets,omitempty"`
+	// PacketsFileName is the file name where the captured packets are temporarily stored in the container. It will be
```
`stored in the container`? I didn't get here, which container do you mean?

hangyan replied:
updated. means the antrea-agent container, but i guess this doesn't need to expose in the api comment.

luolanzone commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -291,3 +291,151 @@ type TLSProtocol struct {
+	Namespace string `json:"namespace,omitempty"`
+	// Pod is the source pod.
+	Pod string `json:"pod,omitempty"`
+	// IP is the source IPv4 or IPv6 address. IP as the source is supported
+	// only for live-traffic Traceflow.
```
`IP as the source is supported only for live-traffic Traceflow`, I suppose this can be removed?

hangyan replied:
done

luolanzone commented on build/charts/antrea/crds/packetcapture.yaml:
I suppose we should update file `docs/api.md` in this PR?

hangyan replied:
added.

antoninbas commented on build/charts/antrea/crds/packetcapture.yaml:
```diff
@@ -0,0 +1,172 @@
+                      oneOf:
+                        - format: ipv4
+                        - format: ipv6
+                packet:
+                  type: object
+                  properties:
```
it looks like we are missing the `srcIP` / `dstIP` properties here?

hangyan replied:
it's actually not missed, since we have a `Source` and `Destination`  field in `spec`, so the `srcIP/dstIP` is not used. The golang structure still keep these fields in the PR, but i think we can removed them. The whole `Packet` structure has changed a lot during the review compared to the original one(==Traceflow's Packet strucuture).

hangyan replied:
Updated this part.  Currently the package related structure looks like this:

```

// Packet includes header info.
type Packet struct {
	IPv6Header      *IPv6Header     `json:"ipv6Header,omitempty"`
	TransportHeader TransportHeader `json:"transportHeader"`
}
```
Note: remove `IPv4Header` as it's unused and in another thread, we are discussing if a `IPFamily` field is needed.

Also in the `TransportHeader`, we have tcp/icmp/udp strcuture to allow user filter based on transport attributes. For icmp, we don't have any filter yet, so the strucutre is mainly used as a type indicator.

```
type ICMPEchoRequestHeader struct {
}
```

hangyan replied:
cc @jianjuns @tnqn @luolanzone 

Can you help review this MR again? Thank you

I will be actively working on this recently.

antoninbas commented on build/charts/antrea/crds/packetcapture.yaml:
```diff
@@ -0,0 +1,172 @@
+    singular: packetcapture
+    kind: PacketCapture
+    shortNames:
+      - pcp
```
do you think `pcap` may be a better name?

hangyan replied:
sounds better!

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -427,3 +354,132 @@ type BGPPeer struct {
+	// Pod is the source pod.
+	Pod string `json:"pod"`
+	// IP is the source IPv4 or IPv6 address.
+	IP *string `json:"ip,omitempty"`
```
you can remove all the string pointers (replace `*string` with `string`).
We had a long discussion about the use of pointers during the `BGPPolicy` API review: https://github.com/antrea-io/antrea/pull/6009#discussion_r1614020224
At least for strings which are introduced as API fields from the get-go(i.e., not added later on), I would avoid pointers. AFAIK we don't have a case where we need to distinguish been "unset" and "empty" string. Here "empty" will always mean "omitted".

hangyan replied:
OK!

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -427,3 +354,132 @@ type BGPPeer struct {
+	// Pod is the source pod.
+	Pod string `json:"pod"`
+	// IP is the source IPv4 or IPv6 address.
+	IP *string `json:"ip,omitempty"`
```
Is this field mutually exclusive with `Namespace` / `Pod`?

hangyan replied:
not exactly. we don't have this kind of check now. This `IP` field is used for  `Pod->IP` or `IP->Pod` filter case.

The only requirement is that at least one pod is specified.

antoninbas replied:
So it would be valid to provide both a destination Pod and a destination IP (which don't match)?

hangyan replied:
As long as there is one pod present(no matter whether it's in the source or destination), it will work(ignore the ip field in the same endpoint). I didn't add strict validation for the IP related case, not sure what's the best approach.

hangyan replied:
since it's exclusive , i have added crd validation using `oneOf` in the yaml. User cannot provide `ip` and `pod` both in the source/dest struct in there CR.

hangyan replied:
@antoninbas found a little problem with this whole `omitempty` thing.

https://github.com/antrea-io/antrea/pull/5821/commits/4ef7a4b15336adb6307e3d952756f4ee9676c25f

Since i'm using `oneOf` to perform some validation in the crd,  so the related field need to be set to `omitempty` wether they are pointer or value or not , or the validation will fail. (using yaml is fine, but golang struct is not, eg in tests code)

For example. If i want to use oneOf for source.ip/pod, means only one can be spcified. the following wont work

```golang
type Source struct {
	Namespace string `json:"namespace"`
	Pod string `json:"pod"`
	IP string `json:"ip"`
}
```
```golang
spec: crdv1alpha1.PacketCaptureSpec{
					Source: crdv1alpha1.Source{
						Namespace: "test-namespace",
						Pod:       "test-pod",
					},
```

since there is no `omitempty` it's actually:
```golang
spec: crdv1alpha1.PacketCaptureSpec{
					Source: crdv1alpha1.Source{
						Namespace: "test-namespace",
						Pod:       "test-pod",
                                                IP:         "",
					},
```

So the `oneOf` validation will fail. The solution would be change every affected field (including `pod`)  to pointer + omitempty. or keep as value + omitempty. If not set, it won't show up. 

Currently i kept the value format and add some of the `omitempty` tag back, as change to pointer would require some changes on the current codebase. If you still felt this is needed ( > pointer), i can update this later.

antoninbas replied:
well I think that in K8s this would be something like:
```go
type Source {
    SourcePod *PodReference `json:"sourcePod,omitempty"`
    SourceIP *string `json:"sourceIP,omitempty"`
}

type Destination {
    DestinationPod *PodReference `json:"destinationPod,omitempty"`
    DestinationService *ServiceReference `json:"destinationService,omitempty"`
    DestinationIP *string `json:"destinationIP,omitempty"`
}
```

Technically the pointers are not strictly necessary for the IP fields, but I like the symmetry.

It does make the API more verbose though, cc @tnqn

antoninbas replied:
now that we have addressed the issue, could you put back the `oneOf`s in the CRD OpenAPI schema?

hangyan replied:
sure ,already did.

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -427,3 +354,132 @@ type BGPPeer struct {
+type IPv6Header struct {
+	// NextHeader is the IPv6 protocol.
+	NextHeader *int32 `json:"nextHeader,omitempty" yaml:"nextHeader,omitempty"`
+	// HopLimit is the IPv6 Hop Limit.
+	HopLimit *int32 `json:"hopLimit,omitempty" yaml:"hopLimit,omitempty"`
```
not sure if this is a useful filter for packet capture?

hangyan replied:
no, will remove this.

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -427,3 +354,132 @@ type BGPPeer struct {
+	Destination   Destination   `json:"destination"`
+	Packet        *Packet       `json:"packet,omitempty"`
+	// FileServer specifies the sftp url config for the fileServer. Captured packets will be uploaded to this server.
+	FileServer BundleFileServer `json:"fileServer"`
```
I feel like we should have the option to write the pcap to a local file as well, just for ease of use / testing
Not everyone will want to have an ftp server for this

hangyan replied:
So do you think keep the files in contains is enough?  Since this is not a cli, it would be odd to export the files to a location on the host.

If keep it in the container, filepath + container name should be ok?

antoninbas replied:
> If keep it in the container, filepath + container name should be ok

Just filepath I think, the container will always be `antrea-agent`? If users want the file to be on the Node, they can provide a path that is mounted from the host, such as `/var/log/antrea`.

But there would need to be a way for users to figure out which antrea-agent Pod has the capture file, so that would need to be included in the Status. Are packets captured at the source or the destination or both?

hangyan replied:
Both could happen, in the source or destination.  pod name can be exposed in the status. Just need to think about the value format for the filepath field.

antoninbas replied:
I think my original point still applies. Always requiring users to provide an SFTP server in order to do a packet capture is not ideal. That being said, if I am the only one who feels strongly about this, we can always address it in the future if needed.

hangyan replied:
Agreed here. I have the following plans:

1. make `FileServer` field optional.
2. Change the `status.packetsFileName` to `status.packetsFilePath` and it's value will be like `antrea-agent-xxxx:/tmp/xxx.pcap

Does this sounds good?

hangyan replied:
see: https://github.com/antrea-io/antrea/pull/6257/commits/939fe064c42d1498353bfd98d8ecc2ba07e38181

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -427,3 +354,132 @@ type BGPPeer struct {
+// UDPHeader describes spec of a UDP header.
+type UDPHeader struct {
+	// SrcPort is the source port.
+	SrcPort *int32 `json:"srcPort,omitempty"`
+	// DstPort is the destination port.
+	DstPort *int32 `json:"dstPort,omitempty"`
+}
+
+// TCPHeader describes spec of a TCP header.
+type TCPHeader struct {
+	// SrcPort is the source port.
+	SrcPort *int32 `json:"srcPort,omitempty"`
+	// DstPort is the destination port.
+	DstPort *int32 `json:"dstPort,omitempty"`
```
I know I suggested pointers for this originally. Given more recent discussions, I don't really feel strongly about it anymore, so given that 0 is not a valid port number typically, I would be ok with removing the pointers. I'll let @tnqn chime in as well.

hangyan replied:
temporary reset to int32 type.

tnqn replied:
My first impression is this should have a pointer because it's optional. `if header.SrcPort != nil` is the condition in which we should capture specifc source port only, right?

hangyan replied:
yes, it's optional. sounds more reasonable to set it as pointer.

hangyan commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -427,3 +354,132 @@ type BGPPeer struct {
+
+// Source describes the source spec of the packetcapture.
+type Source struct {
+	// Namespace is the source namespace.
```
updated

hangyan commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -427,3 +354,132 @@ type BGPPeer struct {
+
+type PacketCapturePhase string
+
+const (
```
also see discussion https://github.com/antrea-io/antrea/pull/6257#discussion_r1600696782

for the constants part i think it's unnecessary,  each CRD may have more different status in the future.

About types, since `Traceflow` is in `vebeta1` now, are you suggesting we define some common types in neutral place?

jianjuns replied:
I am not sure if we should define common types across versions. Let us see if @tnqn and @antoninbas have an opinion.

antoninbas replied:
I guess in theory we could have common definitions in a neutral directory, it wouldn't need to be under `apis/crd/<version>`. But if it's not a versioned directory, then we would have to be confident that they are not going to change.

K8s has `meta/v1` with comm types, such as `Time`. I don't know how good of an analogy it is.

tnqn replied:
Except those meta types in meta/v1. K8s doesn't define common types across versions. If a constant needs to be used by a version, it will just be replicated in that version. I think it's to keep every version's stability and compatability. Otherwise a change in one version would have an impact on other versions.

hangyan commented:
@jianjuns  Do you think we can add a ip family and ip protocol field? so we can remove ipv6header and icmpheader.

jianjuns commented:
> @jianjuns Do you think we can add a ip family and ip protocol field? so we can remove ipv6header and icmpheader.

That works for me. If you do not see a potential requirement to support other IP header fields, esp. v4 or v6 specific fields.

hangyan commented:
> #5821

updated.

hangyan commented:
> > @jianjuns Do you think we can add a ip family and ip protocol field? so we can remove ipv6header and icmpheader.
> 
> That works for me, if you do not see a potential requirement to support other IP header fields, esp. v4 or v6 specific fields. And I meant to use intstr.IntOrString for protocol.

updated. Also this new field need to be align with the `transportHeader` field. Didn't find a good way to perform precheck in the CRD def.

jianjuns commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -427,3 +355,122 @@ type BGPPeer struct {
+type Packet struct {
+	// IPFamily is the filter's ip family . default to `IPv4`.
+	IPFamily v1.IPFamily `json:"ipFamily"`
+	// Protocol represents the transport protocol. default to ICMP(1).
```
We should list what protocol string values we support.

hangyan replied:
you mean in the comments? Is that enough?

hangyan replied:
updated.

jianjuns commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -427,3 +355,122 @@ type BGPPeer struct {
+
+// Packet includes header info.
+type Packet struct {
+	// IPFamily is the filter's ip family . default to `IPv4`.
```
ip -> IP

tnqn commented on build/charts/antrea/crds/packetcapture.yaml:
```diff
@@ -0,0 +1,198 @@
+                  type: integer
+                packetsFileName:
+                  type: string
+
```
remove the blank line

hangyan replied:
done

tnqn commented on build/charts/antrea/crds/packetcapture.yaml:
```diff
@@ -0,0 +1,198 @@
+    kind: PacketCapture
+    shortNames:
+      - pcap
+
```
ditto

tnqn commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -17,81 +17,9 @@ package v1alpha1
+	"k8s.io/apimachinery/pkg/util/intstr"
 )
 
-type TraceflowPhase string
```
could you create another PR to cleanup these dead code?

hangyan replied:
https://github.com/antrea-io/antrea/pull/6667/

created this MR to cleanup the code .will reset the changes in this MR soon.

tnqn commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -427,3 +355,123 @@ type BGPPeer struct {
+
+// TransportHeader describes spec of a TransportHeader.
+type TransportHeader struct {
+	UDP *UDPHeader `json:"udp,omitempty" yaml:"udp,omitempty"`
+	TCP *TCPHeader `json:"tcp,omitempty" yaml:"tcp,omitempty"`
```
remove the yaml tag?

tnqn commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -427,3 +355,123 @@ type BGPPeer struct {
+
+// Packet includes header info.
+type Packet struct {
+	// IPFamily is the filter's IP family . default to `IPv4`.
```
```suggestion
	// IPFamily is the filter's IP family. Default to `IPv4`.
```

hangyan commented:
@antoninbas @jianjuns @tnqn 

Any new comments?

antoninbas commented on build/charts/antrea/crds/packetcapture.yaml:
```diff
@@ -0,0 +1,196 @@
+                      default: IPv4
+                    protocol:
+                      x-kubernetes-int-or-string: true
+                      enum: [ICMP, TCP, UDP, IPv6-ICMP, 1, 6, 17, 58]
```
it feels like for packet capture, we don't need to have a fixed list of protocols? we could filter on an arbitrary protocol value
also SCTP is missing

hangyan replied:
i didn't add support for SCTP yet. will try to work on that. or maybe this is not a must-have feature in this time.

antoninbas replied:
My main point was not about SCTP specifically, but about whether we really need to support a _fixed_ list of protocols, as opposed to supporting a common set of names, but also letting users provide an arbitrary protocol number.

hangyan replied:
sure. remove this for now and i will double check the impl code and add more tests.

hangyan replied:
https://github.com/antrea-io/libOpenflow/blob/f8e71bfb83160a95d2ec6ca53bf55e6f821593e9/protocol/ip.go#L17

Not sure if i missed anything here, but the type code defined in this file doesn't have SCTP.

hangyan commented on multicluster/test/mocks/mock_controller_runtime_manager.go:
reverted.

tnqn commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,123 @@ type BGPPeer struct {
+	NumCapturedPackets *int32 `json:"numCapturedPackets,omitempty"`
+	// PacketsFileName is the file name where the captured packets are temporarily cached. The file will be
+	// removed after the PacketCapture is deleted.
+	PacketsFileName string       `json:"packetsFileName"`
```
What does the file refer to exactly? The file uploaded to fileserver, or the local file in pod?

hangyan replied:
it's a filename for fileserver and also in local container. After the CR is deleted, local file will be deleted too.

Also @antoninbas suggested maybe add container (in which pod) info to the path so we can also support only save the file in the container thus don't require a fileserver. this may also require some `spec` changes(make fileserver optional)

tnqn commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,123 @@ type BGPPeer struct {
+}
+
+type PacketCaptureStatus struct {
+	Phase PacketCapturePhase `json:"phase"`
+	// Reason records the failure reason when the capture fails.
+	Reason string `json:"reason"`
+	// NumCapturedPackets records how many packets have been captured. If it reaches the target number, the capture
+	// can be considered as finished.
+	NumCapturedPackets *int32 `json:"numCapturedPackets,omitempty"`
+	// PacketsFileName is the file name where the captured packets are temporarily cached. The file will be
+	// removed after the PacketCapture is deleted.
+	PacketsFileName string       `json:"packetsFileName"`
+	StartTime       *metav1.Time `json:"startTime,omitempty"`
```
What does StartTime represent and which component is responsible for setting it when the source and destination Pods are on two Nodes?

hangyan replied:
Source. It represents the start time of the capture since we have a `timeout` settings in the spec, so this field can be used to check whether this capture has been timeout. I will update the comments to make this more clear.

tnqn commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,123 @@ type BGPPeer struct {
+	Reason string `json:"reason"`
+	// NumCapturedPackets records how many packets have been captured. If it reaches the target number, the capture
+	// can be considered as finished.
+	NumCapturedPackets *int32 `json:"numCapturedPackets,omitempty"`
```
- When the source and destination Pod are on different Node, will it capture packets on both Nodes? 
- N means the total of the packets on two Nodes or each? 
- What if packets can be captured on source Node and not on destination Node?

hangyan replied:
No. All sessions will only performing  the capture in one node, either in the source or the destination.  I think the location doesn't matter so this was hidden from the api spec.

luolanzone replied:
What do you mean it's "either in the source or the destination", when the user defined both source and destination in the spec, how do we tell which Node the packets are captured from?

tnqn commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,123 @@ type BGPPeer struct {
+}
+
+type PacketCaptureStatus struct {
+	Phase PacketCapturePhase `json:"phase"`
+	// Reason records the failure reason when the capture fails.
+	Reason string `json:"reason"`
```
Not sure a single Phase and Reason are enough for all conditions that it can occur when capturing packets on two Nodes and uploading to a server, maybe it will need a list conditions to represent each condition and respective reason and timestamp. But hard to say what are needed until it's implemented, we can defer to the implementation patch.

hangyan replied:
you mean conditions like `Traceflow`? Currently i didn't implement this in PacketCapture, not sure if this is needed.

tnqn replied:
Phase is a deprecated pattern of K8s API, it recommends to use conditions which is more flexible to express state of each condition. For example, there could be a condition for capture and another condition for upload. See https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties

I don't have strong opinion if the phase and a single reason can cover all possible results.

hangyan replied:
updated. use `conditions` instead of `Phase/Reason`

tnqn reviewed (commented):
I'm curious what the packet number, phase, reason, timestamp mean when the target flow involves two Nodes, and how the two agents coordinate.

tnqn commented:
>```
>type Source {
>    SourcePod *PodReference `json:"sourcePod,omitempty"`
>    SourceIP *string `json:"sourceIP,omitempty"`
>}
>
>type Destination {
>    DestinationPod *PodReference `json:"destinationPod,omitempty"`
>    DestinationService *ServiceReference `json:"destinationService,omitempty"`
>    DestinationIP *string `json:"destinationIP,omitempty"`
>}
>```

@antoninbas I don't find place to reply your comment at https://github.com/antrea-io/antrea/pull/6257#discussion_r1772291362, so starting one here. I get the points of XXXReference and pointer, but is it common to have the struct name in the names of its members? It seems useful when the structs are inline properties in their parents but seems redundant otherwise.

antoninbas commented:
@tnqn 

>  I get the points of XXXReference and pointer, but is it common to have the struct name in the names of its members? It seems useful when the structs are inline properties in their parents but seems redundant otherwise.

No you're right, I am not sure why I added the `Source` / `Destination` prefix in the field names.

hangyan commented:
> I'm curious what the packet number, phase, reason, timestamp mean when the target flow involves two Nodes, and how the two agents coordinate.

If there are two nodes, only the source node is used. Since the user is only cared about the final result, i think use either node is fine.

hangyan commented:
> @tnqn
> 
> > I get the points of XXXReference and pointer, but is it common to have the struct name in the names of its members? It seems useful when the structs are inline properties in their parents but seems redundant otherwise.
> 
> No you're right, I am not sure why I added the `Source` / `Destination` prefix in the field names.

updated. yeah this seems neater.

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,129 @@ type BGPPeer struct {
+	IPFamily v1.IPFamily `json:"ipFamily"`
+	// Protocol represents the transport protocol. default to ICMP(1). Other
+	// possible choices are: TCP(6), UDP(17).
+	Protocol        *intstr.IntOrString `json:"protocol,omitempty"`
```
it feels like the default should be "don't filter on protocol", and not ICMP

hangyan replied:
make sense. updated.

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,129 @@ type BGPPeer struct {
+// Packet includes header info.
+type Packet struct {
+	// IPFamily is the filter's IP family. Default to `IPv4`.
+	IPFamily v1.IPFamily `json:"ipFamily"`
```
nit: maybe add `omitempty`

hangyan replied:
updated.

antoninbas commented on build/charts/antrea/crds/packetcapture.yaml:
```diff
@@ -0,0 +1,206 @@
+                packet:
+                  type: object
+                  x-kubernetes-validations:
+                    - rule: "(self.ipFamily == 'IPv4' && self.protocol != 'IPv6-ICMP' && self.protocol != 58) || (self.ipFamily == 'IPv6' && self.protocol != 'ICMP' && self.protocol != 1) "
```
this doesn't seem correct unless I am missing something. Protocol is not limited to ICMP.

hangyan replied:
removed this validation rule as it seems unnecessary for now.

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,129 @@ type BGPPeer struct {
+	// IPFamily is the filter's IP family. Default to `IPv4`.
+	IPFamily v1.IPFamily `json:"ipFamily"`
+	// Protocol represents the transport protocol. default to ICMP(1). Other
+	// possible choices are: TCP(6), UDP(17).
```
comment no longer up-to-date

hangyan replied:
updated.

antoninbas commented on build/charts/antrea/crds/packetcapture.yaml:
```diff
@@ -0,0 +1,202 @@
+              properties:
+                source:
+                  type: object
+                  nullable: true
```
is the use of nullable correct here. It doesn't seem compatible with the `oneOf` below, unless I am misunderstanding what it means. I also didn't see `nullable` being used in our other CRDs.

hangyan replied:
will double check this.

hangyan replied:
this flag was meant to work with `oneOf`. But  as the fields changes a lot recently,  remove this seems doesn't affect anything now. Removed this for now.

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,128 @@ type BGPPeer struct {
+type Packet struct {
+	// IPFamily is the filter's IP family. Default to `IPv4`.
+	IPFamily v1.IPFamily `json:"ipFamily,omitempty"`
+	// Protocol represents the transport protocol. default is not filter on protocol
```
```suggestion
	// Protocol represents the transport protocol. Default is to not filter on protocol.
```

luolanzone commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,128 @@ type BGPPeer struct {
+
+// Source describes the source spec of the packetcapture.
+type Source struct {
+	// Pod is the source pod,
```
```suggestion
	// Pod is the source Pod.
```

luolanzone commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,128 @@ type BGPPeer struct {
+}
+
+type CaptureConfig struct {
+	FirstN *PacketCaptureFirstNConfig `json:"firstN,omitempty"`
```
why is it 'omitempty'? what is the default value if it can be empty?

hangyan replied:
mainly for other capture options in the future i think.

tnqn replied:
> I think we should keep it as a pointer, and maybe keep the omitempty, but wrap the field in a oneOf? That essentially makes the field required though, and I don't know if it's an ok way to define the API now, while giving us the opportunity to evolve the API later by adding new capture options.

This is what I would suggest too.

antoninbas replied:
@hangyan I don't think this was implemented?

But thinking about it more, I think it is actually ok to omit the `FirstN` field. If it is omitted, the capture should go on until the timeout is reached, and all captured packets should be included in the output file (although we could have an upper bound on the size of the file). That makes `FirstN` truly optional, and `oneOf` would not be appropriate. Maybe that was the intent all along. What do you think @hangyan @tnqn?

By the way this is also why we need a comment for this field.

antoninbas replied:
The the comment should probably read:

> // FirstN means we only capture first N packets from the target traffic.
> // At the moment this is the only supported configuration and it is required for every capture.

hangyan replied:
sure, updated.

luolanzone commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,128 @@ type BGPPeer struct {
+	NumCapturedPackets *int32 `json:"numCapturedPackets,omitempty"`
+	// PacketsFileName is the file name where the captured packets are temporarily cached, also the file name stored in the configured fileserver. The temporary file in the local cache will be removed after the PacketCapture is deleted.
+	PacketsFileName string `json:"packetsFileName"`
+	// StartTime is the time when this capture sessions starts.
```
```suggestion
	// StartTime is the time when this capture session starts.
```

luolanzone commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,129 @@ type BGPPeer struct {
 	GracefulRestartTimeSeconds *int32 `json:"gracefulRestartTimeSeconds,omitempty"`
 }
+
+type PodReference struct {
+	Namespace string
+	Name      string
+}
+
+type ServiceReference struct {
+	Namespace string
+	Name      string
+}
+
```
Do we need two definitions for these? I feel to define one "NamespacedNameReference" or "NameReference" is enough.

hangyan replied:
ok, i prefer separated def too.

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,129 @@ type BGPPeer struct {
+	// NumCapturedPackets records how many packets have been captured. If it reaches the target number, the capture
+	// can be considered as finished.
+	NumCapturedPackets *int32 `json:"numCapturedPackets,omitempty"`
+	// PacketsFilePath is the file path where the captured packets are stored. The format is: "<antrea-agent-pod-name>:<path>".
+	// If `.spec.FileServer` is present, this file will also be uploaded to the targeted location.
+	PacketsFilePath string `json:"packetsFilePath"`
```
I think you originally had in the comment that the file would be deleted when the Custom Resource is deleted? It would be good to add this information back.

hangyan replied:
added

luolanzone commented on build/charts/antrea/crds/packetcapture.yaml:
```diff
@@ -0,0 +1,193 @@
+                      enum: [IPv4, IPv6]
+                      default: IPv4
+                    protocol:
+                      x-kubernetes-int-or-string: true
```
This should be limited to three supported types: udp/tcp/icmp

hangyan replied:
https://github.com/antrea-io/antrea/pull/6257#discussion_r1772285151

there is a discussion about this before. I personally think maybe SCTP is not needed for now.

antoninbas replied:
@luolanzone there is no reason to limit protocols here in my opinion. We are capturing traffic, not generating traffic, and Pod traffic can use any protocol. Only Services are limited to TCP / UDP / SCTP.
We could restrict protocol strings to what we "know" (as we need to do the mapping), but protocol numbers should be arbitrary.
If there is a way to restrict the string value to known values, while allowing arbitrary numeric values, then we can go for it.

hangyan replied:
In the code MR we are planning to remove the support for service. As we don't have a good way to distinguish the traffic between service and normal pod -> pod. I think we can keep the current format, and at least add some tests for SCTP.  Using  tcpdump filter to perform the capture(we are switching to gopacket/libpcap from ovs), it should work on any protocol value.

antoninbas replied:
I don't think that's the point I was trying to make.
I was trying to convey that K8s itself doesn't limit what transport protocols can be used for Pod-to-Pod traffic, and so this feature shouldn't restrict it either by forcing the protocol to be one of TCP, UDP, ICMP, SCTP. This is why we should support an arbitrary protocol number here, even though we should also support some pre-defined strings for convenience.
There is no need to test SCTP support specifically. From an Antrea perspective, there is no real difference in the implementation between protocols. It's just a number that we use to define a packet capture filter.

hangyan replied:
sure !

hangyan commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,129 @@ type BGPPeer struct {
 	GracefulRestartTimeSeconds *int32 `json:"gracefulRestartTimeSeconds,omitempty"`
 }
+
+type PodReference struct {
+	Namespace string `json:"namespace,omitempty"`
+	Name      string `json:"name,omitempty"`
+}
+
+type ServiceReference struct {
+	Namespace string `json:"namespace,omitempty"`
+	Name      string `json:"name,omitempty"`
+}
```
sure, will update this. I think i copied this from some `ObjectReference` from `corev1`, not sure why they do that.

antoninbas replied:
I think it's for legacy / historical reason. If you look at this type for example, you will see that both fields are required: https://github.com/kubernetes/api/blob/a432ebea600276f14fbf74e8a5186063d4a4216a/admissionregistration/v1/types.go#L1189-L1195

(but I am not sure what the best approach is for namespace, required or optional)

hangyan commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,129 @@ type BGPPeer struct {
+	Destination   Destination   `json:"destination"`
+	Packet        *Packet       `json:"packet,omitempty"`
+	// FileServer specifies the sftp url config for the fileServer. Captured packets will be uploaded to this server.
+	FileServer *BundleFileServer `json:"fileServer,omitempty"`
```
sure

tnqn commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,137 @@ type BGPPeer struct {
+}
+
+type PacketCaptureSpec struct {
+	Timeout       *uint16       `json:"timeout,omitempty"`
+	CaptureConfig CaptureConfig `json:"captureConfig"`
+	Source        Source        `json:"source"`
+	Destination   Destination   `json:"destination"`
+	Packet        *Packet       `json:"packet,omitempty"`
+	// FileServer specifies the sftp url config for a file server. If present, captured packets will be uploaded to this server.
+	FileServer *PacketCaptureFileServer `json:"fileServer,omitempty"`
```
The spec doesn't provide enough information to understand how to use the API. I have the questions when looking at the fields:
1. What's the behavior if Timeout is nil, capturing forever or there is a default value?
2. Could I set Source only or Destination only, or they must be set together? If not the latter, why they are not optional like other fields?
3. I suppose it means capturing all traffic matching Source and Destination when Packet is nil, but please make this explicit in the comment.
4. If FileServer is present, captured packets will be uploaded to this server, then what if FileServer is not present? This seems a valid input given it's a pointer.

hangyan replied:
sure, will add more comments.

hangyan replied:
updated

tnqn commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,137 @@ type BGPPeer struct {
+	Reason string `json:"reason"`
+	// NumCapturedPackets records how many packets have been captured. If it reaches the target number, the capture
+	// can be considered as finished.
+	NumCapturedPackets *int32 `json:"numCapturedPackets,omitempty"`
```
I feel this doesn't need to be a pointer. It's correct to use `NumCapturedPackets=0` to represent nothing has been captured.

hangyan replied:
ok

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,144 @@ type BGPPeer struct {
+
+// Packet includes header info.
+type Packet struct {
+	// IPFamily is the filter's IP family. Default to `IPv4`.
```
```suggestion
	// IPFamily is the filter's IP family. Defaults to IPv4.
```

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,144 @@ type BGPPeer struct {
+}
+
+type PacketCaptureSpec struct {
+	// Timeout is the timeout for this capture session. If not specified, default to 60s.
```
defaults to 60s

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,144 @@ type BGPPeer struct {
+	Number int32 `json:"number"`
+}
+
+const DefaultPacketCaptureTimeout uint16 = 60
```
is this used anywhere?

hangyan replied:
in the datapath code, or we can move it there.

hangyan replied:
removed.

hangyan commented on build/charts/antrea/crds/packetcapture.yaml:
```diff
@@ -0,0 +1,189 @@
+          name: Destination-Pod
+          type: string
+          priority: 10
+        - jsonPath: .spec.destination.ip
+          description: The IP address of the destination.
+          name: Destination-IP
+          type: string
+          priority: 10
```
added

antoninbas commented on build/charts/antrea/crds/packetcapture.yaml:
```diff
@@ -0,0 +1,194 @@
+          description: The phase of the PacketCapture.
+          name: Phase
+          type: string
+        - jsonPath: .spec.source.pod
+          description: The name of the source Pod.
+          name: Source-Pod
+          type: string
+          priority: 10
+        - jsonPath: .spec.destination.pod
+          description: The name of the destination Pod.
+          name: Destination-Pod
+          type: string
+          priority: 10
+        - jsonPath: .spec.source.ip
+          description: The IP address of the source.
+          name: Source-IP
+          type: string
+          priority: 10
+        - jsonPath: .spec.destination.ip
+          description: The IP address of the destination.
+          name: Destination-IP
+          type: string
+          priority: 10
```
This is not great: showing the Pod name without the namespace is not accurate enough, and we don't handle the destination Service case.

I don't recall what is possible to do with `jsonPath` here, but if the expression language is rich enough, maybe we could just have a `source` and `destination` columns, and populate them correctly based on what was provided by the user.

For example, for `source` we would want the JSONPath expression to evaluate to:
- `.spec.source.pod.namespace + "/" + .spec.source.pod.name` if `.spec.source.pod` is not null
- `.spec.source.ip` otherwise

Not sure if this is supported, but the current printer columns are not ideal IMO.

hangyan replied:
I think it's fine for now. Since `Pod` is a struct, so the results will be like:

```
$ kubectl get pcap -o wide -A
NAME      PHASE     SOURCE-POD                                  DESTINATION-POD                            SOURCE-IP   DESTINATION-IP   TIMEOUT   AGE
pc-test   Running   {"name":"frontend","namespace":"default"}   {"name":"backend","namespace":"default"}                                20        7s
```

For service case as discussed in the other PR before, we don't have a perfect solution for now ( the previous code only work for some cases), so it's removed.

antoninbas replied:
Ok, I missed that we were printing the struct

Hopefully this can be improved in the future

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,146 @@ type BGPPeer struct {
+
+// Destination describes the destination spec of the PacketCapture.
+type Destination struct {
+	// Pod is the destination Pod, exclusive with destination Service.
```
you seem to have removed Service from this API version, so the comment is no longer valid
However, it is mutually exclusive with IP

We should also say that IP and Pod are mutually exclusive for the Source struct above

hangyan replied:
ok

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,146 @@ type BGPPeer struct {
+	// all kinds of traffic will count.
+	Packet *Packet `json:"packet,omitempty"`
+	// FileServer specifies the sftp url config for a file server. If present, captured packets will be uploaded to this server.
+	// If not, packets file will only be present in the antrea-agent pod.
```
```suggestion
	// If not, the packet capture results will only be present as a file in the antrea-agent container.
```

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,146 @@ type BGPPeer struct {
+	Packet *Packet `json:"packet,omitempty"`
+	// FileServer specifies the sftp url config for a file server. If present, captured packets will be uploaded to this server.
+	// If not, packets file will only be present in the antrea-agent pod.
+	// When capture finished, the path information will be shown in `.status.PacketsFilePath`.
```
When the capture finishes

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,146 @@ type BGPPeer struct {
+	// can be considered as finished.
+	NumCapturedPackets int32 `json:"numCapturedPackets,omitempty"`
+	// PacketsFilePath is the file path where the captured packets are stored. The format is: "<antrea-agent-pod-name>:<path>".
+	// If `.spec.FileServer` is present, this file will also be uploaded to the targeted location. This file
```
s/targeted location/target location

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,146 @@ type BGPPeer struct {
+	// If `.spec.FileServer` is present, this file will also be uploaded to the targeted location. This file
+	// will be removed after the PacketCapture CR is deleted.
+	PacketsFilePath string `json:"packetsFilePath"`
+	// StartTime is the time when this capture session starts.
```
s/starts/started

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,146 @@ type BGPPeer struct {
+
+const (
+
+	// CaptureCompleted means enough packets has been captured and saved in antrea-agent Pod locally already, but hasn't been
```
s/has been capture/have been captured
s/but hasn't been/but results haven't been

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,146 @@ type BGPPeer struct {
+const (
+
+	// CaptureCompleted means enough packets has been captured and saved in antrea-agent Pod locally already, but hasn't been
+	// uploaded yet(if a fileserver has been configured).
```
missing space before parenthesis

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,146 @@ type BGPPeer struct {
+type PacketCaptureConditionType string
+
+const (
+
```
remove extra empty line

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,146 @@ type BGPPeer struct {
+
+	// CaptureCompleted means enough packets has been captured and saved in antrea-agent Pod locally already, but hasn't been
+	// uploaded yet(if a fileserver has been configured).
+	CaptureCompleted PacketCaptureConditionType = "CaptureCompleted"
+
+	// PacketsUploaded means the captured packets file has been uploaded to the target fileserver.
+	PacketsUploaded PacketCaptureConditionType = "PacketsUploaded"
```
you need a more uniform naming scheme for your conditions. See https://pkg.go.dev/k8s.io/api/core/v1#NodeConditionType

maybe `PacketCaptureCompleted` and `PacketCaptureFileUploaded`

we may need additional conditions based on other discussions to indicate that a capture is pending (because of too many concurrent captures) or that a capture is ongoing but has not completed yet

hangyan replied:
OK, updated.

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,148 @@ type BGPPeer struct {
+
+// Source describes the source spec of the packetcapture.
+type Source struct {
+	// Pod is the source Pod, exclusive with source `IP`.
```
```suggestion
	// Pod is the source Pod, mutually exclusive with IP.
```

not sure why you used backticks specifically here

hangyan replied:
updated.

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,148 @@ type BGPPeer struct {
+
+// Destination describes the destination spec of the PacketCapture.
+type Destination struct {
+	// Pod is the destination Pod, exclusive with destination `IP`.
```
ditto

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,148 @@ type BGPPeer struct {
+
+const (
+	// PacketCapturePending means this request is still pending.
+	PacketCapturePending PacketCaptureConditionType = "PacketCapturePedning"
```
typo: s/PacketCapturePedning/PacketCapturePending

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,148 @@ type BGPPeer struct {
+	PacketCapturePending PacketCaptureConditionType = "PacketCapturePedning"
+	// PacketCaptureRunning means antrea is processing this capture request.
+	PacketCaptureRunning PacketCaptureConditionType = "PacketCaptureRunning"
+	// CaptureCompleted means enough packets has been captured and saved in antrea-agent Pod locally already, but results hasn't been
```
```suggestion
	// CaptureCompleted means enough packets have been captured and saved in antrea-agent Pod locally already, but results haven't been
```

I think I highlighted this in a previous round of review already

hangyan replied:
OK, will remember this.

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,148 @@ type BGPPeer struct {
+}
+
+// PacketCaptureFileServer specifies the PacketCapture file server information.
+type PacketCaptureFileServer struct {
+	// The URL of the file server. It is set with format: scheme://host[:port][/path],
+	// e.g, https://api.example.com:8443/v1/supportbundles/. If scheme is not set, https is used by default.
+	URL string `json:"url"`
+}
```
As I pointed out in the other PR, it seems that we don't have an equivalent to `BundleServerAuthConfiguration` for PacketCapture (to configure sftp authentication). But that may have been intentional?

hangyan replied:
yes. a dedicated  secret is serving the purpose.

antoninbas reviewed (dismissed):
LGTM, it would be good to get a second approval

luolanzone commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,148 @@ type BGPPeer struct {
+	IP *string `json:"ip,omitempty"`
+}
+
+// TransportHeader describes spec of a TransportHeader.
+type TransportHeader struct {
+	UDP *UDPHeader `json:"udp,omitempty"`
+	TCP *TCPHeader `json:"tcp,omitempty"`
+}
+
+// UDPHeader describes spec of a UDP header.
+type UDPHeader struct {
+	// SrcPort is the source port.
+	SrcPort *int32 `json:"srcPort,omitempty"`
+	// DstPort is the destination port.
+	DstPort *int32 `json:"dstPort,omitempty"`
+}
+
+// TCPHeader describes spec of a TCP header.
```
spec of -> the spec of

luolanzone commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,148 @@ type BGPPeer struct {
+type Packet struct {
+	// IPFamily is the filter's IP family. Defaults to IPv4.
+	IPFamily v1.IPFamily `json:"ipFamily,omitempty"`
+	// Protocol represents the transport protocol. Default is to not filter on protocol.
```
```suggestion
	// Protocol represents the transport protocol. No protocol based filter when it's empty.
```

luolanzone commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,148 @@ type BGPPeer struct {
+// PacketCaptureFileServer specifies the PacketCapture file server information.
+type PacketCaptureFileServer struct {
+	// The URL of the file server. It is set with format: scheme://host[:port][/path],
+	// e.g, https://api.example.com:8443/v1/supportbundles/. If scheme is not set, https is used by default.
```
```suggestion
	// e.g., https://api.example.com:8443/v1/supportbundles/. If scheme is not set, https is used by default.
```

luolanzone commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,148 @@ type BGPPeer struct {
+type PacketCaptureStatus struct {
+	// NumCapturedPackets records how many packets have been captured. If it reaches the target number, the capture
+	// can be considered as finished.
+	NumCapturedPackets int32 `json:"numCapturedPackets,omitempty"`
```
I suppose `capturedPacketsNum` is preferred? @antoninbas @tnqn

tnqn replied:
Similar fields in K8s API is "NumberXXXed", I would prefer "NumberCaptured", no need to have "packet" as there is no ambiguity.
https://github.com/kubernetes/kubernetes/blob/a3ea994fa2810395287870db56630345d96a5fe7/staging/src/k8s.io/api/apps/v1/types.go#L681

And I think it shouldn't have `omitempty`, as 0 should be shown as 0, not to be omitted.

hangyan replied:
OK, updated.

luolanzone commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,148 @@ type BGPPeer struct {
+	PacketsFilePath string `json:"packetsFilePath"`
+	// StartTime is the time when this capture session started.
+	StartTime *metav1.Time `json:"startTime,omitempty"`
+	// Condition represents the latest available observations of the PacketCapture 's current state.
```
```suggestion
	// Condition represents the latest available observations of the PacketCapture's current state.
```

luolanzone commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,148 @@ type BGPPeer struct {
+	PacketCapturePending PacketCaptureConditionType = "PacketCapturePending"
+	// PacketCaptureRunning means antrea is processing this capture request.
+	PacketCaptureRunning PacketCaptureConditionType = "PacketCaptureRunning"
+	// CaptureCompleted means enough packets have been captured and saved in antrea-agent Pod locally already, but results haven't been
```
```suggestion
	// PacketCaptureCompleted means enough packets have been captured and saved in an antrea-agent Pod locally already, but results haven't been
```

luolanzone commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,148 @@ type BGPPeer struct {
+	// PacketCaptureRunning means antrea is processing this capture request.
+	PacketCaptureRunning PacketCaptureConditionType = "PacketCaptureRunning"
+	// CaptureCompleted means enough packets have been captured and saved in antrea-agent Pod locally already, but results haven't been
+	// uploaded yet (if a fileserver has been configured).
+	PacketCaptureCompleted PacketCaptureConditionType = "PacketCaptureCompleted"
+	// PacketsUploaded means the captured packets file has been uploaded to the target fileserver.
```
```suggestion
	// uploaded yet (if a file server has been configured).
	PacketCaptureCompleted PacketCaptureConditionType = "PacketCaptureCompleted"
	// PacketCaptureFileUploaded means the captured packets file has been uploaded to the target file server.
```

tnqn commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,148 @@ type BGPPeer struct {
+	// Timeout is the timeout for this capture session. If not specified, defaults to 60s.
+	Timeout       *uint16       `json:"timeout,omitempty"`
+	CaptureConfig CaptureConfig `json:"captureConfig"`
+	// Source is the traffic source we want to perform capture on. Both `Source` and `Destination` is reuqired
```
```suggestion
	// Source is the traffic source we want to perform capture on. Both `Source` and `Destination` is required
```

tnqn commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,148 @@ type BGPPeer struct {
+	Timeout       *uint16       `json:"timeout,omitempty"`
+	CaptureConfig CaptureConfig `json:"captureConfig"`
+	// Source is the traffic source we want to perform capture on. Both `Source` and `Destination` is reuqired
+	// for a capture session, and at least one `Pod` should present either in the source or the destination.
```
```suggestion
	// for a capture session, and at least one `Pod` should be present either in the source or the destination.
```

tnqn commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,148 @@ type BGPPeer struct {
+	// for a capture session, and at least one `Pod` should present either in the source or the destination.
+	Source      Source      `json:"source"`
+	Destination Destination `json:"destination"`
+	// Packet defines what kind of traffic we want to capture between the source and desination. If not spicified,
```
```suggestion
	// Packet defines what kind of traffic we want to capture between the source and destination. If not specified,
```
You could turn on typo check in your IDE, there are quite a few typos.

hangyan replied:
Ok i thought `make spelling-fix` can help auto correct this.

tnqn commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,148 @@ type BGPPeer struct {
+	// can be considered as finished.
+	NumCapturedPackets int32 `json:"numCapturedPackets,omitempty"`
+	// PacketsFilePath is the file path where the captured packets are stored. The format is: "<antrea-agent-pod-name>:<path>".
+	// If `.spec.FileServer` is present, this file will also be uploaded to the target location. This file
+	// will be removed after the PacketCapture CR is deleted.
+	PacketsFilePath string `json:"packetsFilePath"`
```
I feel we shoud use this path for storing the URL path when `FileServer` is specified. And perhaps just name it `FilePath` to correspond to `FileServer`

tnqn commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,148 @@ type BGPPeer struct {
+	// If `.spec.FileServer` is present, this file will also be uploaded to the target location. This file
+	// will be removed after the PacketCapture CR is deleted.
+	PacketsFilePath string `json:"packetsFilePath"`
+	// StartTime is the time when this capture session started.
+	StartTime *metav1.Time `json:"startTime,omitempty"`
```
Since there are conditions, CaptureStart can be a condition, and its LastTransitionTime will be the start time, and we don't need another time here.

hangyan commented:
@luolanzone All updated. Can you take a look again?

I also removed the `ipv6` related support in CRD schema. It should be easy to add it back in the following releases.

tnqn commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -354,3 +355,146 @@ type BGPPeer struct {
+	// NumberCaptured records how many packets have been captured. If it reaches the target number, the capture
+	// can be considered as finished.
+	NumberCaptured int32 `json:"numberCaptured"`
+	// FilePath is the file path where the captured packets are stored. The format is: "<antrea-agent-pod-name>:<path>".
+	// If `.spec.FileServer` is present, this file will also be uploaded to the target location. This file
+	// will be removed after the PacketCapture CR is deleted.
```
Consider this explanation:

FilePath specifies the location where captured packets are stored. It can either be a URL to download the pcap file (if "Spec.FileServer" is specified) or a local file path on the antrea-agent Pod where the packet was captured, formatted as <antrea-agent-pod-name>:<path>. When using a local file path, the file will be automatically removed after the PacketCapture resource is deleted.

tnqn commented on build/yamls/antrea.yml:
```diff
@@ -2896,6 +2896,205 @@ spec:
+              properties:
+                startTime:
+                  type: string
+                numCapturedPackets:
```
needs update

tnqn commented on build/yamls/antrea.yml:
```diff
@@ -2896,6 +2896,205 @@ spec:
+                  type: string
+                numCapturedPackets:
+                  type: integer
+                packetsFilePath:
```
ditto

tnqn commented on build/yamls/antrea.yml:
```diff
@@ -2896,6 +2896,205 @@ spec:
+            status:
+              type: object
+              properties:
+                startTime:
```
ditto

tnqn reviewed (approved):
LGTM

luolanzone reviewed (approved):
LGTM

antoninbas commented:
/skip-all
